src/main/java/bio/terra/cli/Context.java

```diff
@@ -0,0 +1,207 @@
+package bio.terra.cli;
+
+import bio.terra.cli.exception.SystemException;
+import bio.terra.cli.exception.UserActionableException;
+import bio.terra.cli.serialization.disk.DiskContext;
+import bio.terra.cli.utils.JacksonMapper;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Internal representation of the current context or state. This class maintains singleton instances
+ * of the internal state classes (Config, Server, User, Workspace).
+ */
+public class Context {
+  private static final Logger logger = LoggerFactory.getLogger(Context.class);
+
+  // singleton objects that define the current context or state
+  private static Config currentConfig;
+  private static Server currentServer;
+  private static User currentUser;
+  private static Workspace currentWorkspace;
+
+  // env var name to optionally override where the context is persisted on disk
+  private static final String CONTEXT_DIR_OVERRIDE_NAME = "TERRA_CONTEXT_PARENT_DIR";
+
+  // file paths related to persisting the context on disk
+  private static final String CONTEXT_DIRNAME = ".terra";
+  private static final String CONTEXT_FILENAME = "context.json";
+  private static final String PET_KEYS_DIRNAME = "pet-keys";
+  private static final String LOGS_DIRNAME = "logs";
+  private static final String LOG_FILENAME = "terra.log";
+
+  /**
+   * Reads the context file from disk and initializes the singleton internal state classes (Config,
+   * Server, User, Workspace).
+   */
+  public static void initializeFromDisk() {
+    try {
+      // try to read in an instance of the context file
+      DiskContext diskContext =
+          JacksonMapper.readFileIntoJavaObject(getContextFile().toFile(), DiskContext.class);
+      currentConfig = new Config(diskContext.config);
+      currentServer = new Server(diskContext.server);
+      currentUser = diskContext.user == null ? null : new User(diskContext.user);
+      currentWorkspace =
+          diskContext.workspace == null ? null : new Workspace(diskContext.workspace);
+    } catch (IOException ioEx) {
+      // file not found is a common error here (e.g. first time running the CLI, there will be no
+      // pre-existing global context file). we handle this by returning an object populated with
+      // default values below. so, no need to log or throw the exception returned here.
```

Comment:
Ah looks like my warning comment on this method got dropped. Will add it back.

We can't do any logging in this initialize method, because we need to get the logging levels from the config before setting up the loggers. So if we log here, it uses default settings and not the ones specified by the user.